### ISSUE TITLE
The ttnn.linear function outputs an error function when bias exists in tt_transformers

### ISSUE THREAD

**Issue by velonica0** [tenstorrent/tt-metal#25935]
### Describe the bug

When I was experimenting with various LLM modules, I found that when dealing with bias, using the bias parameter of ttnn.linear could not produce correct results, and ordinary addition was required.
<img width="2560" height="1380" alt="Image" src="https://github.com/user-attachments/assets/80fc1003-2666-4dc3-91d1-3e34077fb46b" />

In the models/tt_transformers/tt/attention.py file, I see the FIXME comment. What is the reason for this?
<img width="2560" height="1380" alt="Image" src="https://github.com/user-attachments/assets/7cc2892e-1ce0-4a2d-aeed-fd612a837dbc" />

**Comment by milank94** [comment]
@sraizada-tt would you be able to provide any insight?

**Comment by sraizada-tt** [comment]
ttnn.linear has a bug with the bias addition 
@yieldthought do you have any info?

**Comment by yieldthought** [comment]
I raised this with the ttnn team [internally](https://tenstorrent.slack.com/archives/C086ECGJ532/p1738838998612099?thread_ts=1738776548.717889&cid=C086ECGJ532) back when I ran into it. Summary of my learnings:
- `ttnn.linear` with bias add works without tracing
- `ttnn.linear` expects `bias shape = [1, 1, 1, out_features]`, but `as_tensor` returns `[1, 1, 32, out_features]` even if you give it a 1D tensor when using `TILE_LAYOUT`
- `ttnn.reshape` can be used to view the produced bias as a (1, 1, 1, dim) tensor if and only if it is a multidevice tensor, otherwise it asserts

I did not have time to produce a minimal reproducer at the time (was working on a P0) and it looks like there was no follow-up.

**Comment by bbradelTT** [comment]
@velonica0 @yieldthought would one of you be able to provide a minimal reproducer? This looks like an issue between as_tensor, reshape, and linear, and it'll be challenging to split out the various effects.

**Comment by bbradelTT** [comment]
@yieldthought the behaviour changes depending on whether the add is fused within the matmul op or not. If the add is fused, the functionality seems to work. When the add is not fused, then the behaviour depends on ttnn.add. With the introduction if binary_ng, an error is thrown in your scenario. The following is a minimal pytest I used to try to reproduce the behaviour. It might be worthwhile to update that comment/code.

```
import pytest
import torch
import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc
from models.utility_functions import torch_random, is_close, comp_allclose


def test_mm4(device):
    torch.manual_seed(0)
    pcc = 0.99
    torch_input_tensor_a = torch.ones([32, 1, 64, 96])
    torch_input_tensor_b = torch.ones([1, 1, 96, 64])
    torch_input_tensor_c = 100*torch.ones([1, 1, 1, 64])
    torch_input_tensor_c2 = 100*torch.ones([1, 1, 32, 64])
    torch_output_tensor = torch.matmul(torch_input_tensor_a, torch_input_tensor_b) + torch_input_tensor_c

    input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
    input_tensor_b = ttnn.from_torch(torch_input_tensor_b, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
    # Basing on torch_input_tensor_c would make the test pass.
    input_tensor_c = ttnn.from_torch(torch_input_tensor_c2, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)

    # Passing in core_grid currently leads to fusing the add
    output1 = ttnn.linear(input_tensor_a, input_tensor_b, bias=input_tensor_c, core_grid=ttnn.CoreGrid(y=8,x=7))
    output_tensor = ttnn.to_torch(output1)
    assert_with_pcc(torch_output_tensor, output_tensor, pcc=pcc)
    print(f"Output with fused add within matmul\n OUTPUT1 {output1}\nTORCH {torch_output_tensor}")
    # Not passing in core_grid currently leads to a composite op
    # Will fail in the add with an "Invalid subtile broadcast type" error.
    output2 = ttnn.linear(input_tensor_a, input_tensor_b, bias=input_tensor_c)
    output_tensor = ttnn.to_torch(output2)
    assert_with_pcc(torch_output_tensor, output_tensor, pcc=pcc)
    print(f"Output with separate add after matmul\n OUTPUT2 {output2}\nTORCH {torch_output_tensor}")
```

@velonica0 could you please provide a concrete example, similar to the above, of linear not producing correct results?

**Comment by velonica0** [comment]
Sorry @bbradelTT 
I'd like to provide you with a simple example, but my code currently has a bug.

[bias.py](https://github.com/user-attachments/files/21521709/bias.txt)

What puzzles me is that when I do ttnn.linear in tt_transformer, the variable is 
memory_layout = <TensorMemoryLayout.WIDTH_SHARDED: 3>, but there is no error.

<img width="2560" height="1380" alt="Image" src="https://github.com/user-attachments/assets/cf107b60-eef6-4acc-a21c-e0cf2e4ac394" />

<img width="2560" height="1380" alt="Image" src="https://github.com/user-attachments/assets/da33f065-cfae-4e42-9f5a-3029cb7f14b3" />

Could you please help me with this and two very minor issues in my code?
Thank you very much for your help.

**Comment by bbradelTT** [comment]
@velonica0 

When there is no program config passed in to matmul/linear, an algorithm is used to determine what to use. That algorithm currently has many limitations when it comes to the more advanced combinations of input sharding and what is allowed with the various optimized backend versions of matmul. Improving this has been lower priority because most people use interleaved tensors, or at most shard the first input tensor.

The model code specifies their own program config to not run into this issue. I think it'd be a hand tuned MatmulMultiCoreReuseMultiCast1DProgramConfig with mcast_in0.

You'll need to either
- mimic what they're doing or
- try with interleaved tensors, possibly moving on to have the first tensor sharded.

**Comment by fahmedchanTT** [comment]
@velonica0 Have you tried either of @bbradelTT 's suggestions ?

**Comment by fahmedchanTT** [comment]
Closed as resolved. @velonica0 please re-open if necessary.
